server/src/main/java/org/apache/druid/segment/realtime/appenderator/StreamAppenderator.java

```diff
@@ -1098,37 +1109,59 @@ SinkSchemaAnnouncer getSinkSchemaAnnouncer()
   {
-    segmentAnnouncer.unannounceSegment(baseSegment);
+    synchronized (sink) {
+      final SegmentIdWithShardSpec baseId = SegmentIdWithShardSpec.fromDataSegment(baseSegment);
+      if (!baseSegmentToUpgradedSegments.containsKey(baseId)) {
+        return;
+      }
 
-    final Set<SegmentIdWithShardSpec> upgradedVersionsOfSegment
-        = baseSegmentToUpgradedVersions.remove(baseSegment.getId());
-    if (upgradedVersionsOfSegment == null || upgradedVersionsOfSegment.isEmpty()) {
-      return;
+      final List<SegmentIdWithShardSpec> upgradedVersionsOfSegment
+          = ImmutableList.copyOf(baseSegmentToUpgradedSegments.get(baseId));
+
+      for (SegmentIdWithShardSpec newId : upgradedVersionsOfSegment) {
+        final DataSegment newSegment = new DataSegment(
+            newId.getDataSource(),
+            newId.getInterval(),
+            newId.getVersion(),
+            baseSegment.getLoadSpec(),
+            baseSegment.getDimensions(),
+            baseSegment.getMetrics(),
+            newId.getShardSpec(),
+            baseSegment.getBinaryVersion(),
+            baseSegment.getSize()
+        );
+        unannounceSegment(newSegment);
+      }
     }
+  }
 
-    for (SegmentIdWithShardSpec newId : upgradedVersionsOfSegment) {
-      final DataSegment newSegment = new DataSegment(
-          newId.getDataSource(),
-          newId.getInterval(),
-          newId.getVersion(),
-          baseSegment.getLoadSpec(),
-          baseSegment.getDimensions(),
-          baseSegment.getMetrics(),
-          newId.getShardSpec(),
-          baseSegment.getBinaryVersion(),
-          baseSegment.getSize()
-      );
-      segmentAnnouncer.unannounceSegment(newSegment);
+  private void unannounceSegment(DataSegment segment)
+  {
+    try {
+      segmentAnnouncer.unannounceSegment(segment);
+    }
+    catch (Exception e) {
+      log.makeAlert(e, "Failed to unannounce segment[%s]", schema.getDataSource())
+         .addData("identifier", segment.getId().toString())
+         .emit();
+    }
+    SegmentIdWithShardSpec id = SegmentIdWithShardSpec.fromDataSegment(segment);
+    SegmentIdWithShardSpec baseId = upgradedSegmentToBaseSegment.remove(id);
+    if (baseId == null) {
+      return;
+    }
+    baseSegmentToUpgradedSegments.get(baseId).remove(id);
+    if (baseSegmentToUpgradedSegments.get(baseId).isEmpty()) {
+      baseSegmentToUpgradedSegments.remove(baseId);
     }
   }
 
-  public void registerNewVersionOfPendingSegment(
-      SegmentIdWithShardSpec basePendingSegment,
-      SegmentIdWithShardSpec newSegmentVersion
-  ) throws IOException
+  public void registerNewVersionOfPendingSegment(PendingSegmentRecord pendingSegmentRecord) throws IOException
   {
+    SegmentIdWithShardSpec basePendingSegment = idToPendingSegment.get(pendingSegmentRecord.getUpgradedFromSegmentId());
+    SegmentIdWithShardSpec newSegmentVersion = pendingSegmentRecord.getId();
```

Comment:
Rename for homogeneity with rest of the Druid code:
```suggestion
    SegmentIdWithShardSpec upgradedPendingSegment = pendingSegmentRecord.getId();
```

Reply:
Done. Renamed the method as well